modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItIndexAndIndexStorageDestructionTest.java

```diff
@@ -0,0 +1,157 @@
+            IndexStorage indexStorage = schemaAwareStorage.storage();
+
+            if (indexStorage instanceof AbstractPageMemoryIndexStorage) {
+                ((AbstractPageMemoryIndexStorage<?, ?>) indexStorage).transitionToDestroyedState();
+            } else if (indexStorage instanceof AbstractRocksDbIndexStorage) {
+                ((AbstractRocksDbIndexStorage) indexStorage).transitionToDestroyedState();
+            } else {
+                fail("Index storage is not of a type supported by the test: " + indexStorage.getClass());
+            }
+        }
+    }
+
+    private int indexId(String indexName) {
+        CatalogIndexDescriptor indexDescriptor = node.catalogManager().aliveIndex(indexName, node.clock().nowLong());
+        assertThat(indexDescriptor, is(notNullValue()));
+
+        return indexDescriptor.id();
+    }
+
+    @Test
+    void insertTouchingDestroyedIndexStorageDoesNotFailWholeWrite() {
+        KeyValueView<Integer, String> view = node.tables().table(TABLE_NAME)
+                .keyValueView(Integer.class, String.class);
+
+        assertDoesNotThrow(() -> view.put(null, ANOTHER_KEY, "Mary"));
+
+        assertThat(view.get(null, ANOTHER_KEY), is("Mary"));
+    }
+
+    @Test
+    void updateTouchingDestroyedIndexStorageDoesNotFailWholeWrite() {
+        KeyValueView<Integer, String> view = node.tables().table(TABLE_NAME)
+                .keyValueView(Integer.class, String.class);
+
+        assertDoesNotThrow(() -> view.put(null, PREEXISTING_KEY, "Mary"));
+
+        assertThat(view.get(null, PREEXISTING_KEY), is("Mary"));
+    }
+
+    @Test
+    void replaceTouchingDestroyedIndexStorageDoesNotFailWholeWrite() {
+        KeyValueView<Integer, String> view = node.tables().table(TABLE_NAME)
+                .keyValueView(Integer.class, String.class);
+
+        assertDoesNotThrow(() -> view.replace(null, PREEXISTING_KEY, "John", "Mary"));
+
+        assertThat(view.get(null, PREEXISTING_KEY), is("Mary"));
+    }
+
+    @Test
+    void sqlReadFromDestroyedIndexStorageFailsWithStalePlanError() {
+        SqlException ex = assertThrows(
+                SqlException.class,
+                () -> cluster.query(0, "SELECT * FROM " + TABLE_NAME + " WHERE name = 'John'", rs -> null)
+        );
+
+        assertThat(ex.code(), is(STALE_PLAN_ERR));
+        assertThat(ex.getMessage(), is("The plan is stale. Please retry."));
+    }
+
+    @Test
+    void kvQueryFromDestroyedIndexStorageFailsWithStalePlanError() {
+        KeyValueView<Tuple, Tuple> view = node.tables().table(TABLE_NAME).keyValueView();
+
+        CriteriaException ex = assertThrows(CriteriaException.class, () -> {
+            try (Cursor<?> cursor = view.query(null, null, INDEX_NAME)) {
+                cursor.next();
+            }
+        });
+
+        assertThat(ex.code(), is(STALE_PLAN_ERR));
+        assertThat(ex.getMessage(), is("The plan is stale. Please retry."));
```

Comment:
This error message looks strange. Who is it intended for?

Reply:
For the user who issued the query. What's strange in it?

Reply:
Changed the error message. I will ask guys from the SQL team to take a look; if they are ok, I'll as the doc writers.